File: scheduler/core/src/main/java/io/orkes/conductor/dao/scheduler/SchedulerDAO.java
package io.orkes.conductor.dao.scheduler;

import java.util.List;
import java.util.Map;
import java.util.Set;

import com.netflix.conductor.common.run.SearchResult;

import io.orkes.conductor.scheduler.model.WorkflowScheduleExecutionModel;
import io.orkes.conductor.scheduler.model.WorkflowScheduleModel;

public interface SchedulerDAO {

    void updateSchedule(WorkflowScheduleModel workflowSchedule);

    void saveExecutionRecord(WorkflowScheduleExecutionModel executionModel);

    WorkflowScheduleExecutionModel readExecutionRecord(String executionId);

    void removeExecutionRecord(String executionId);

    WorkflowScheduleModel findScheduleByName(String name);

    List<WorkflowScheduleModel> findAllSchedules(String workflowName);

    void deleteWorkflowSchedule(String name);

    List<String> getPendingExecutionRecordIds();

    List<WorkflowScheduleModel> getAllSchedules();

    Map<String, WorkflowScheduleModel> findAllByNames(Set<String> workflowScheduleNames);

    long getNextRunTimeInEpoch(String scheduleName);

    void setNextRunTimeInEpoch(String name, long toEpochMilli);

    SearchResult<WorkflowScheduleModel> searchSchedules(
            String workflowName,
            String scheduleName,
            Boolean paused,
            String freeText,
            int start,
            int size,
            List<String> sortOptions);
}

File: scheduler/core/src/main/typescript/io/orkes/conductor/dao/scheduler/SchedulerDAO.ts
import { SearchResult } from 'com/netflix/conductor/common/run/SearchResult';
import { WorkflowScheduleExecutionModel } from 'io/orkes/conductor/scheduler/model/WorkflowScheduleExecutionModel';
import { WorkflowScheduleModel } from 'io/orkes/conductor/scheduler/model/WorkflowScheduleModel';

export interface SchedulerDAO {
    updateSchedule(workflowSchedule: WorkflowScheduleModel): void;

    saveExecutionRecord(executionModel: WorkflowScheduleExecutionModel): void;

    readExecutionRecord(executionId: string): WorkflowScheduleExecutionModel | null;

    removeExecutionRecord(executionId: string): void;

    findScheduleByName(name: string): WorkflowScheduleModel | null;

    findAllSchedules(workflowName: string): Array<WorkflowScheduleModel>;

    deleteWorkflowSchedule(name: string): void;

    getPendingExecutionRecordIds(): Array<string>;

    getAllSchedules(): Array<WorkflowScheduleModel>;

    findAllByNames(workflowScheduleNames: Set<string>): Map<string, WorkflowScheduleModel>;

    getNextRunTimeInEpoch(scheduleName: string): number;

    setNextRunTimeInEpoch(name: string, toEpochMilli: number): void;

    searchSchedules(
        workflowName: string | null,
        scheduleName: string | null,
        paused: boolean | null,
        freeText: string | null,
        start: number,
        size: number,
        sortOptions: Array<string> | null
    ): SearchResult<WorkflowScheduleModel>;
}
